src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,532 @@
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            writeToolsVersions();
+
+            WriteByteArrayValue(writer, "publicKey", publicKey.AsSpan());
+            writer.WriteKey("options");
+            WriteCompilationOptions(writer, compilationOptions);
+
+            writer.WriteKey("syntaxTrees");
+            writer.WriteArrayStart();
+            foreach (var syntaxTree in syntaxTrees)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                WriteSyntaxTree(writer, syntaxTree, pathMap, options, cancellationToken);
+            }
+            writer.WriteArrayEnd();
+
+            writer.WriteKey("references");
+            writer.WriteArrayStart();
+            foreach (var reference in references)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                WriteMetadataReference(writer, reference, pathMap, options, cancellationToken);
+            }
+            writer.WriteArrayEnd();
+            writer.WriteObjectEnd();
+
+            void writeToolsVersions()
+            {
+                writer.WriteKey("toolsVersions");
+                writer.WriteObjectStart();
+                if ((options & DeterministicKeyOptions.IgnoreToolVersions) == 0)
+                {
+                    var compilerVersion = typeof(Compilation).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                    writer.Write("compilerVersion", compilerVersion);
+
+                    var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                    writer.Write("runtimeVersion", runtimeVersion);
+
+                    writer.Write("frameworkDescription", RuntimeInformation.FrameworkDescription);
+                    writer.Write("osDescription", RuntimeInformation.OSDescription);
+                }
+
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteSyntaxTree(
+            JsonWriter writer,
+            SyntaxTreeKey syntaxTree,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            WriteFilePath(writer, "fileName", syntaxTree.FilePath, pathMap, options);
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText(cancellationToken));
+            writer.WriteKey("parseOptions");
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
```

Comment:
Should we do `writer.WriteNull();` here (similar to how `WriteType` handles nulls)?

Reply:
Good catch that was a bug. Added a test to expose the bug and fixed this up.